crates/bevy_animation/src/spline_group.rs

```diff
@@ -0,0 +1,186 @@
+use splines::Spline;
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum LoopStyle {
+    Once,
+    Loop,
+    PingPong,
+}
+
+pub trait SplineGroup {
+    type Sample;
+
+    fn splines(&self) -> Vec<&Spline<f32, f32>>;
+
+    fn loop_style(&self) -> LoopStyle;
+    fn loop_style_mut(&mut self) -> &mut LoopStyle;
+
+    fn time(&self) -> f32;
+    fn time_mut(&mut self) -> &mut f32;
+
+    fn speed(&self) -> f32;
+    fn speed_mut(&mut self) -> &mut f32;
+
+    fn paused(&self) -> bool;
+    fn paused_mut(&mut self) -> &mut bool;
+
+    fn pong(&self) -> bool;
+    fn pong_mut(&mut self) -> &mut bool;
+
+    fn sample(&self, time: f32) -> Self::Sample;
+
+    fn current(&self) -> Self::Sample {
+        self.sample(self.time())
+    }
+
+    fn is_empty(&self) -> bool {
+        if self.splines().is_empty() {
+            true
+        } else {
+            self.splines()
+                .into_iter()
+                .fold(true, |acc, v| if !v.is_empty() { false } else { acc })
+        }
+    }
+
+    fn start_time(&self) -> Option<f32> {
+        let starts: Vec<f32> = self
+            .splines()
+            .into_iter()
+            .map(|s| spline_start_time(s))
+            .filter_map(|s| s)
+            .collect();
+
+        if starts.is_empty() {
+            None
+        } else {
+            Some(
+                starts
+                    .iter()
+                    .fold(starts[0], |acc, v| if *v < acc { *v } else { acc }),
+            )
+        }
+    }
+
+    fn end_time(&self) -> Option<f32> {
+        let ends: Vec<f32> = self
+            .splines()
+            .into_iter()
+            .map(|s| spline_end_time(s))
+            .filter_map(|s| s)
+            .collect();
+
+        if ends.is_empty() {
+            None
+        } else {
+            Some(
+                ends.iter()
+                    .fold(ends[0], |acc, v| if *v > acc { *v } else { acc }),
+            )
+        }
+    }
+
+    fn duration(&self) -> Option<f32> {
+        let start = self.start_time();
+        let end = self.end_time();
+        if start.is_none() || end.is_none() {
+            None
+        } else {
+            Some((start.unwrap() - end.unwrap()).abs())
+        }
+    }
+
+    fn advance(&mut self, delta_time: f32) {
+        if self.is_empty() || self.paused() {
+            return;
+        }
+
+        let start = self.start_time().unwrap();
+        let end = self.end_time().unwrap();
+        let direction = self.speed().signum();
+        let reversed = direction < 0.0;
+        let past_boundary = if reversed {
+            if self.pong() {
+                end < self.time()
+            } else {
+                start > self.time()
+            }
+        } else {
+            if self.pong() {
+                start > self.time()
+            } else {
+                end < self.time()
+            }
+        };
```

Comment:
Clippy is complaining about the collapsible-if on CI
https://github.com/bevyengine/bevy/runs/1106650581#step:6:249
```suggestion
        } else if self.pong() {
            start > self.time()
        } else {
            end < self.time()
        };
```